docs/rfc/rfc-026-p2p-bad-peers-checktx.md

```diff
@@ -0,0 +1,375 @@
+seen for the first time. Afterwards, the  transaction is cached, to avoid running `CheckTx` on transactions already checked. 
+Thus when a transaction is received from a peer, if it is in the cache,
+`CheckTx` is not ran again, but the peers' ID is addded to the list of peers who sent this particular transaction.
+These transactions are rechecked once a block is committed to verify that they are still valid. 
+
+If invalid transactions are not kept in the cache, they can be resubmitted multiple times, and `CheckTx` will be executed on them upon submission therefore we do not need to remember the previous response codes for these transactions.
+
+Currently there are two ways to implement peer banning based on the result of `CheckTx`:
+
+1. Introduce banning when transactions start to be processed
+2. Adapt the recheck logic to support this
+
+
+**Peer banning when transactions are received**
+
+If a transaction fails `CheckTx` the [first time it is seen](https://github.com/tendermint/tendermint/blob/ff0f98892f24aac11e46aeff2b6d2c0ad816701a/mempool/v0/clist_mempool.go#L409),  the peer can be banned right there:
+
+>>mempool/v0/clist_mempool.go#L409
+```golang
+
+      if (r.CheckTx.Code == abci.CodeTypeOK) && postCheckErr == nil {
+          // Check Tx passed
+      } else {
+			// ignore bad transaction
+          mem.logger.Debug(
+            "rejected bad transaction",
+            "tx", types.Tx(tx).Hash(),
+            "peerID", peerP2PID,
+            "res", r,
+            "err", postCheckErr,
+          )
+          mem.metrics.FailedTxs.Add(1)
+
+          mem.banPeer(peerP2PID)
+
+          if !mem.config.KeepInvalidTxsInCache {
+            // remove from cache (it might be good later)
+            mem.cache.Remove(tx)
+          } else {
+            // If transactins stay in the cache, remember they failed
+            mem.cache.invalidCachedTx.Store(tx.Key(), true)
+          }
+		}
+
+```					
+The `KeepInvalidTxsInCache` configuration parameter defines whether an invalid transaction stays in cache. 
+Currently, even transactions that never could have been valid are kept in the cache. This can be mitigated by 
+expanding the condititon in the code above with:
+
+```golang
+if !mem.config.KeepInvalidTxsInCache || r.CheckTx.Code == abci.NeverValid {
+            // remove from cache (it might be good later)
+    mem.cache.Remove(tx)
+}
+```
+
+As said, this code will [never be executed](https://github.com/tendermint/tendermint/blob/ff0f98892f24aac11e46aeff2b6d2c0ad816701a/mempool/v0/clist_mempool.go#L239) for transactions whose signature is found
+in the cache. 
+Instead of remembering the cached transactions, we could have had a valid/invalit bit per transaction within the cache. As transactions themselves do not
```

Comment:
```suggestion
Instead of remembering the cached transactions, we could have had a valid/invalid bit per transaction within the cache. As transactions themselves do not
```